src/docker.rs

```diff
@@ -0,0 +1,190 @@
+            Ok(res) => {
+                if res.status.success() {
+                    info!("Successfully created base Image");
+                    Ok(())
+                } else {
+                    error!("Failed to build base image");
+                    Err(String::from_utf8_lossy(&res.stderr).into())
+                }
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    fn create_base_image(
+        &self,
+        solana_root_path: &Path,
+        image_name: String,
+        docker_path: &PathBuf,
+        validator_type: &ValidatorType,
+    ) -> Result<Output, Box<dyn Error>> {
+        self.create_dockerfile(validator_type, docker_path, None)?;
+
+        trace!("Tmp: {}", docker_path.as_path().display());
+        trace!("Exists: {}", docker_path.as_path().exists());
+
+        // We use std::process::Command here because Docker-rs is very slow building dockerfiles
+        // when they are in large repos. Docker-rs doesn't seem to support the `--file` flag natively.
+        // so we result to using std::process::Command
+        let dockerfile = docker_path.join("Dockerfile");
+        let context_path = solana_root_path.display().to_string();
+
+        let progress_bar = new_spinner_progress_bar();
+        progress_bar.set_message(format!("{BUILD}Building {validator_type} docker image...",));
+
+        let command = format!(
+            "docker build -t {}/{image_name}:{} -f {dockerfile:?} {context_path}",
+            self.registry, self.tag,
+        );
+
+        let output = match Command::new("sh")
+            .arg("-c")
+            .arg(&command)
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .spawn()
+            .expect("Failed to execute command")
+            .wait_with_output()
+        {
+            Ok(res) => Ok(res),
+            Err(err) => Err(Box::new(err) as Box<dyn Error>),
+        };
+        progress_bar.finish_and_clear();
+        info!("{validator_type} image build complete");
+
+        output
+    }
+
+    fn copy_file_to_docker(
+        source_dir: &Path,
+        docker_dir: &Path,
+        file_name: &str,
+    ) -> std::io::Result<()> {
+        let source_path = source_dir.join("src/scripts").join(file_name);
+        let destination_path = docker_dir.join(file_name);
+        fs::copy(source_path, destination_path)?;
+        Ok(())
+    }
+
+    fn create_dockerfile(
+        &self,
+        validator_type: &ValidatorType,
+        docker_path: &PathBuf,
+        content: Option<&str>,
+    ) -> Result<(), Box<dyn Error>> {
+        if docker_path.exists() {
+            fs::remove_dir_all(docker_path)?;
+        }
+        fs::create_dir_all(docker_path)?;
+
+        if let DeployMethod::Local(_) = self.deploy_method {
+            if validator_type == &ValidatorType::Bootstrap {
+                let manifest_path =
+                    PathBuf::from(env::var("CARGO_MANIFEST_DIR").expect("$CARGO_MANIFEST_DIR"));
```

Comment:
I just did a little test, and `CARGO_MANIFEST_DIR` is only set when invoking the program using `cargo`, and not just calling the executable directly.

Since this might be slightly brittle, would it make sense to also pass a path to the scripts directory?

Reply:
what is the best way to do this? Just add a flag that is something like: `--startup-scripts-dir <path>`? or were you thinking something else?

Reply:
we also use this here: https://github.com/anza-xyz/validator-lab/blob/044c9d2ae25ab99a98022398a82244d99e885b72/src/lib.rs#L19-L21. We need this for the `release-channel` deployment method. So maybe we just pass in a `--validator-lab-dir <absolute-path-to-validator-lab-directory>`. From here we can get startup script directory and then of course the directory needed for `release-channel` deployment

Reply:
That makes sense to me!

Reply:
cool! it's implemented in: https://github.com/anza-xyz/validator-lab/pull/8